bycycle/cyclepoints.py

```diff
@@ -173,66 +140,64 @@ def find_zerox(sig, ps, ts):
-    zerox_rise = np.zeros(n_rises, dtype=int)
-    for idx_rise in range(n_rises):
-        sig_temp = np.copy(sig[ts[idx_rise]:ps[idx_rise + 1 - idx_bias] + 1])
-        sig_temp -= (sig_temp[0] + sig_temp[-1]) / 2.
+        n_decays -= 1
+        idx_bias += 1
 
-        # If data is all zeros, just set the zero-crossing to be halfway between
-        if np.sum(np.abs(sig_temp)) == 0:
-            zerox_rise[idx_rise] = ts[idx_rise] + int(len(sig_temp) / 2.)
+    rises = _find_zerox_flanks(sig, 'rise', n_rises, troughs, peaks, idx_bias)
+    decays = _find_zerox_flanks(sig, 'decay', n_decays, peaks, troughs, idx_bias)
 
-        # If rise is actually decay, just set the zero-crossing to be halfway between
-        elif sig_temp[0] > sig_temp[-1]:
-            zerox_rise[idx_rise] = ts[idx_rise] + int(len(sig_temp) / 2.)
+    return rises, decays
 
-        else:
-            zerox_rise[idx_rise] = ts[idx_rise] + int(np.median(_fzerorise(sig_temp)))
 
-    # Find zero-crossings for decays
-    zerox_decay = np.zeros(n_decays, dtype=int)
-    for idx_decay in range(n_decays):
-        sig_temp = np.copy(sig[ps[idx_decay]:ts[idx_decay + idx_bias] + 1])
+def _find_zerox_flanks(sig, flank, n_flanks, extrema_start, extrema_end, idx_bias):
+    """Helper function for find_zerox."""
+
+    assert flank in ['rise', 'decay']
+    idx_bias = -idx_bias + 1 if flank == 'rise' else idx_bias
+    comp = gt if flank == 'rise' else lt
+
+    flanks = np.zeros(n_flanks, dtype=int)
+    for idx in range(n_flanks):
+
+        sig_temp = np.copy(sig[extrema_start[idx]:extrema_end[idx + idx_bias] + 1])
         sig_temp -= (sig_temp[0] + sig_temp[-1]) / 2.
 
         # If data is all zeros, just set the zero-crossing to be halfway between
         if np.sum(np.abs(sig_temp)) == 0:
-            zerox_decay[idx_decay] = ps[idx_decay] + int(len(sig_temp) / 2.)
+            flanks[idx] = extrema_start[idx] + int(len(sig_temp) / 2.)
+
+        # If flank is actually an extrema, just set the zero-crossing to be halfway between
+        elif comp(sig_temp[0], sig_temp[-1]):
+            flanks[idx] = extrema_start[idx] + int(len(sig_temp) / 2.)
 
-        # If decay is actually rise, just set the zero-crossing to be halfway between
-        elif sig_temp[0] < sig_temp[-1]:
-            zerox_decay[idx_decay] = ps[idx_decay] + int(len(sig_temp) / 2.)
         else:
-            zerox_decay[idx_decay] = ps[idx_decay] + int(np.median(_fzerofall(sig_temp)))
+            flanks[idx] = extrema_start[idx] + int(np.median(find_flank_zero_xs(sig_temp, flank)))
 
-    return zerox_rise, zerox_decay
+    return flanks
 
 
-def extrema_interpolated_phase(sig, ps, ts, zerox_rise=None, zerox_decay=None):
-    """Use peaks (phase 0) and troughs (phase pi/-pi) to estimate instantaneous phase.
-    Also use rise and decay zero-crossings (phase -pi/2 and pi/2, respectively) if provided.
+def extrema_interpolated_phase(sig, peaks, troughs, rises=None, decays=None):
```

Comment:
I stared at this function for a while, and I feel like there is a refactor available for it, I just didn't follow through on how. But I think it might be worth figuring out the logic of the doubled up code here, and at least document why there are two, and perhaps refactor.

Reply:
I refactor this (3c42063). I don't think there is any duplicate code here. pha_tpi and pha_npi have inverted signs and are merged together. This is so the returned phase is always rising. The function I added does the merging.